Adding fftw

psteinb opened the pull request:
added float32 and float64 support for fftw, I based the implementation of fftw on cufft
there are still a couple of issues:

1) shall we make the number of threads a runtime configurable variable (right now I am simply using all cores available), if we want to do this, the question is how!
2) estimating the plan size does not work with fftw by design (I return `0 bytes` for now)
3) I am using FFTW_MEASURE for now, we could make this a runtime configurable as well or a simple compile time flag.

Commit:
added cmake find module for fftw3

Commit:
used cufft header as templated for fftw

Commit:
added minimal support for findFFTW

Commit:
added components to FindFFTW

Commit:
added options to select openmp/pthreads fftw

Commit:
finished setting up fftw build in Cmake

Commit:
added fftw section

Commit:
finished implementation of serial fftw launch

Commit:
updated and refactor FindFFTW cmake module

Commit:
linker flags are treated correctly now

Commit:
Merge branch 'master' into adding_fftw

Commit:
removed obsolete cmake message statements

Commit:
Merge remote-tracking branch 'upstream/master' into adding_fftw

Commit:
added multi-threading using all cores available for now

tdd11235813 commented:
cool, nice work, this looks great. I want to do some small tests before merging then...

According to number of threads: To allow user defined configuration, I prefer a variable as program option. However, the class for program options itself should be not cluttered with lib-dependent switches (eg. `if(fftw) then add_option(number_threads)`). There are now 2 ways:
1. client code has ability to inject own program options in addition to default options set
   - `gearshifft_fftw` would have different extra program options then `gearshifft_clfft`
2. program option is used everywhere, no matter if needed
   - Help message hints that this or that program option might be ignored.

Since, I already have the program option for device architecture (atm it only relates to OpenCL), 
I would go with the second way for now.  I want to focus on gathering and evaluating results in the next weeks.
What do you think?

tdd11235813 commented on inc/fftw.hpp:
```diff
@@ -0,0 +1,629 @@
+	n_ = std::accumulate(extents_.begin(),
+			     extents_.end(),
+			     1,
+			     std::multiplies<unsigned>());
```
unsigned -> size_t (comes from my older branch, sed did not get this one)

tdd11235813 commented on inc/fftw.hpp:
```diff
@@ -0,0 +1,629 @@
+       * Returns estimated allocated memory on device for FFT plan, accessing the plan size in fftw is impossible
+       
+       */
+      size_t getPlanSize() {
```
getPlanSize() should return 0 (it seems there is only a meta data structure whose size can be neglected here)
Sizes of transform buffers are already returned by getAllocSize().

tdd11235813 commented on inc/fftw.hpp:
```diff
@@ -0,0 +1,629 @@
+      Extent extents_;
+
+      PlanType			plan_           ;
+      value_type*		data_           = nullptr;
```
Being curious, do you prefer `value_type` over `RealOrComplexType` ?

psteinb replied:
yes, as it's more clear for the client of your API

tdd11235813 commented on cmake/FindFFTW.cmake:
```diff
@@ -0,0 +1,273 @@
+  if(EXISTS ENV{FFTW_ROOT})
+    set( FFTW_ROOT $ENV{FFTW_ROOT} )
+  endif()
+endif()
```
did not work out of the box, had to change to (cmake 2.8):

```
if( NOT FFTW_ROOT AND (DEFINED ENV{FFTWDIR} OR DEFINED ENV{FFTW_ROOT}))
  if(EXISTS "$ENV{FFTWDIR}/")                                          
    set( FFTW_ROOT $ENV{FFTWDIR} )                                     
  endif()                                                              
  if( EXISTS "$ENV{FFTW_ROOT}/" )                                      
    set( FFTW_ROOT $ENV{FFTW_ROOT} )                                   
  endif()                                                              
endif()                                                                
```

psteinb replied:
thanks for testing that, I'll add it

tdd11235813 commented on inc/fftw.hpp:
```diff
@@ -0,0 +1,629 @@
+
+      template<typename THostData>
+      void upload(THostData* input) {
+
```
input contains the example data, which should be loaded into the fftw owned array.
input probably could be reused, then copy in download() is not necessary.
In the other case we could benchmark fftw malloc. So, we have two benchmark cases for fftw now.
I would prefer the own fftw malloc'd array (might be proper aligned for SSE/AVX), and in the statistics we could separate the timings, what do you think?

psteinb replied:
ups, I forgot about that, I'll fill upload/download right away

tdd11235813 commented on inc/fftw.hpp:
```diff
@@ -0,0 +1,629 @@
+
+      template<typename THostData>
+      void download(THostData* output) {
+
```
copy computed data to output.
(If nothing happens to output and input pointer (which are actually the same), then there are no errors, thus Boost test is fine.)

psteinb replied:
did that!

tdd11235813 commented on inc/fftw.hpp:
```diff
@@ -0,0 +1,629 @@
+
+      ~FftwImpl(){
+	traits::thread_api<TPrecision>::cleanup_threads();
+      }
```
can you move these thread setup and cleanup to Context as you already mentioned in ToDo there?
fftw thread environment will be set up before first benchmark and cleaned up after last benchmark (Context is instantiated gearshifft life-time).

psteinb replied:
I thought about this for a long time and am still not quite sure which solution is better. I'll move it to the context for now.

psteinb replied:
looking at it in more detail, I have to say that I can't move the thread init calls to the context as it requires the precision to be known. fftw is so "well" written that initializing threads depends on the type to be used!! aka there is fftwf_init_threads for float32_t and there is fftw_init_threads for float64_t data! we can also make the context be TPrecision aware but that may result in more refactoring in the other FFT benchmarks. what do you think?

tdd11235813 replied:
oh now I see, stupid. Benchmarks are changing precision types, while (global) context is initialized and destroyed only one time and should stay precision independent. So, we have to live with it for now. I will look into this later, when next refactoring of backend is pending.

Commit:
applied patches based on comments by tdd11235813

psteinb commented:
I haven't merged https://github.com/mpicbg-scicomp/gearshifft/pull/29 yet! should I do so before we merge this one?

tdd11235813 commented:
yeah, so the getTransferSize() method needs to be implemented in your fftw class...

Commit:
Merge remote-tracking branch 'origin/master' into adding_fftw

psteinb commented:
no problem, I am working on it.

psteinb commented:
I am having fun with the fftw type model:

```
std::complex<double>* input = ...
fftw_complex* data_ = ...
std::copy(input, input + size,data_);
```

will not work, as `fftw_complex` is effectively a `double [2]` and arrays are not assignable. So contrary to my initial assumption, the problem is not Real2D (which should still be std::complex IMHO).

Commit:
circumventing the poor fftw type model with a memcpy
(oh dear creator of the universe, please forgive me as I have used c99 in c++14 code)

tdd11235813 commented:
well, if fftw_complex is double[2], then we can cast it to an equal structured type:
`std::copy(input, input+size, reinterpret_cast<std::complex<double>*>(data))`
You are right, std::complex is sufficient and should be used instead of that exotic Real2D, which I never really liked ;) Thanks for the hint!

Commit:
* copy std::size-t to int for fftw planning functions
* use forward and backward plans for fftw setup
* refactor using allocated_extents and extents for more clear code
  structure
* tested with 1D,2D and 3D transforms

psteinb commented:
fftw extension done from my side, please review

tdd11235813 commented:
thanks a lot, I am testing right now... so std::copy still did not work with the cast?

psteinb commented:
nope

tdd11235813 commented:
for a later optimization: 
what about `FFTW_ESTIMATE` instead of `FFTW_MEASURE`?
- would not overwrite buffers and we can separate the plan creation as wanted
- can speedup up if heuristic makes a good hit, otherwise can slow down, but how much?
  - we need to test of course

for this PR: can we move these two lines to init_forward()?

```
fwd_plan_ = traits::plan<TPrecision>::create(extents_, data_, data_transform_, traits::forward); 
bwd_plan_ = traits::plan<TPrecision>::create(extents_, data_transform_, data_, traits::backward);
```

I would like to separate the memory allocation and the plan creation for the time measurements.
Of course, init_forward() does not sound that it creates a backward plan.
But problem with putting it into malloc() is, that I could not figure out, why malloc() took so long.
It turns out that almost 100% of the runtime in my example came from creating these two plans using FFTW_MEASURE.

Commit:
moved plan creation into init_forward to make timings consistent

psteinb commented:
I'd stick with FFTW_MEASURE for now. From my own experience, FFTW_ESTIMATE can easily produce plans that are 2-4x slower than otherwise. We wouldn't play fair, if we'd use it. But we may consider making it a compile-time or even runtime flag if possible.

I moved the plan create around as you suggested, we can still split it, but it'll produce equally unfair timings as it does now. thinking of which, that would gain us though a correct timing for creating the forward plan.

tdd11235813 commented:
yeah we still do not know, which plan creation took longer, the forward or the backward one.
But that is a later topic, for now I am happy to have fftw on board :) Thanks a bunch for your work!
